packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts

```diff
@@ -106,16 +152,58 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
   }
 
   public get parseNode(): NodeParser {
-    return (row: { [columnName: string]: any }): SourceInstanceHierarchyNode => {
+    return async (row: { [columnName: string]: any }): Promise<SourceInstanceHierarchyNode> => {
       const hasFilterTargetAncestor: boolean = !!row[ECSQL_COLUMN_NAME_HasFilterTargetAncestor];
       const isFilterTarget: boolean = !!row[ECSQL_COLUMN_NAME_IsFilterTarget];
       const filterTargetOptions: HierarchyFilteringPathOptions | undefined = row[ECSQL_COLUMN_NAME_FilterTargetOptions]
         ? JSON.parse(row[ECSQL_COLUMN_NAME_FilterTargetOptions])
         : undefined;
-      const filteredChildrenIdentifierPaths: HierarchyFilteringPath[] | undefined = row[ECSQL_COLUMN_NAME_FilteredChildrenPaths]
-        ? JSON.parse(row[ECSQL_COLUMN_NAME_FilteredChildrenPaths])
-        : undefined;
-      const defaultNode = (this._source.parseNode ?? defaultNodesParser)(row);
+
+      // istanbul ignore next
+      const allFilterPathsIdentifierPositions: Array<[number, number]> | undefined = row[ECSQL_COLUMN_NAME_FilterECInstanceId]
+        ? (this._pathsIdentifierPositions.get({
+            id: `0x${row[ECSQL_COLUMN_NAME_FilterECInstanceId].toString(16)}`,
+            className: "",
+            imodelKey: this._imodelAccess.imodelKey,
+          }) ?? [])
+        : [];
+
+      // _pathsIdentifierPositions doesn't care about classNames, we need to check if identifiers with the same id are
+      // of the same class / derived class / is derived from class
+      const className: string = row[ECSQL_COLUMN_NAME_FilterClassName] ?? "";
+      let filteredChildrenIdentifierPaths: HierarchyFilteringPath[] | undefined;
+      for (const [pathIndex, identifierIndex] of allFilterPathsIdentifierPositions) {
+        // istanbul ignore if
+        if (this._nodeIdentifierPaths.length <= pathIndex) {
+          continue;
+        }
+        const path = HierarchyFilteringPath.normalize(this._nodeIdentifierPaths[pathIndex]);
+        // istanbul ignore if
+        if (path.path.length <= identifierIndex) {
+          continue;
+        }
```

Comment:
I'd like tests to cover these two edge cases.

Reply:
These two edge cases should never happen because:
We create `_pathsIdentifierPositions` array based on `_nodeIdentifierPaths`. So `_pathsIdentifierPositions` always has indexes pointing to an existing place in the `_nodeIdentifierPaths` array.
These two checks were added just a precaution. Removed them.